File: 백준/Gold/7576. 토마토/토마토.py
import sys
from collections import deque

W, H = map(int, sys.stdin.readline().strip().split())

tomato = [[0] * W for _ in range(H)]
visited = [[False] * W for _ in range(H)]

for i in range(H):
    tomato[i] = list(map(int, sys.stdin.readline().strip().split()))

temp = 1

def bfs():
    global visited, q
    # 상 우 하 좌
    dx = [0, 1, 0, -1]
    dy = [-1, 0, 1 ,0]
    while q:
        y, x = q.popleft()
        for i in range(4):
            idx = x + dx[i]
            idy = y + dy[i]
            if idx >= 0 and idy >= 0 and idx < W and idy < H:
                if tomato[idy][idx] == 0 and not visited[idy][idx]:
                    q.append((idy,idx))
                    visited[idy][idx] = True
                    tomato[idy][idx] = tomato[y][x] + 1

q = deque()

for i in range(H):
    for j in range(W):
        if tomato[i][j] == 1:
            q.append((i,j))
            visited[i][j] = True

bfs()
count = 0
for i in range(H):
    for j in range(W):
        if tomato[i][j] == 0:
            count = -1
if count != -1:
    count = max(map(max, tomato)) - 1
print(count)

File: 백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;


public class Main {

    static int W;
    static int H;
    static int[][] map;
    static int[][] result;
    static Queue<int[]> q = new LinkedList<>();
    // 상 우 하 좌
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {-1, 0, 1, 0};
    static int max = 0;


    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

        String[] s = bf.readLine().split(" ");
        W = Integer.parseInt(s[0]);
        H = Integer.parseInt(s[1]);

        map = new int[H][W];
        result = new int[H][W];

        for (int i = 0; i < H; i++) {
            s = bf.readLine().split(" ");
            for (int j = 0; j < W; j++) {
                map[i][j] = Integer.parseInt(s[j]);
                if(Integer.parseInt(s[j]) == -1) result[i][j] = Integer.parseInt(s[j]);
            }
        }

        for (int i = 0; i < H; i++) {
            for (int j = 0; j < W; j++) {
                if (map[i][j] == 1) {
                    result[i][j] = -1;
                    q.add(new int[]{i, j, 0});
                }
            }
        }

        bfs();

        int zeroCount = 0;
        for(int i=0; i<H; i++){
            for(int j=0; j<W; j++){
                if(result[i][j] == 0){
                    zeroCount++;
                }
            }
        }

        if(zeroCount != 0) System.out.println(-1);
        else System.out.println(max);
    }

    public static void bfs() {
        while (!q.isEmpty()) {
            int[] temp = q.poll();
            int y = temp[0];
            int x = temp[1];
            int count = temp[2];
            for (int a = 0; a < 4; a++) {
                int ndx = x + dx[a];
                int ndy = y + dy[a];
                if (ndx >= 0 && ndy >= 0 && ndx < W && ndy < H) {
                    if (map[ndy][ndx] == 0 && result[ndy][ndx] == 0) {
                        q.add(new int[]{ndy, ndx, count + 1});
                        if(max < count + 1) max = count + 1;
                        map[ndy][ndx] = 1;
                        result[ndy][ndx] = count + 1;
                    }
                }
            }
        }
    }


}